### ISSUE TITLE
Requests to AnkiConnect timeout when creating media-heavy cards

### ISSUE THREAD

**Issue by jqhoogland** [Pseudonium/Obsidian_to_Anki#139]
First off, thanks for this awesome plugin.

The problem: I can easily and quickly make a few dozen new cards from a file using the Q-A custom regex.
But as soon as I start adding pictures (already at as few as 5) the requests timeout.

E.g.: The note on the left never finishes converting while the note on the right finishes within seconds.
![image](https://user-images.githubusercontent.com/28595964/102013522-ff9ba300-3d50-11eb-8fc1-c19c99e15239.png)

My idea for a temporary fix was to add one or two new notes at a time, scan the vault, then move on. I figured if the media were already transferred, it wouldn't be able to slow down future requests. But that didn't work.

![image](https://user-images.githubusercontent.com/28595964/102013785-8604b480-3d52-11eb-8ef6-bc3e357d116a.png)

With a quick glance at the callback above, my best guess is that the issue lies in `FileManager.requests_1`, specifically:
``` js
...
  requests.push(multi(temp));
  temp = [];
  console.log("Requesting addition of media...");
  for (let file of this.ownFiles) { 
    const mediaLinks = difference(file.formatter.detectedMedia, this.added_media_set);
    for (let mediaLink of mediaLinks) {
      console.log("Adding media file: ", mediaLink);
      this.added_media_set.add(mediaLink);
      const dataFile = this.app.metadataCache.getFirstLinkpathDest(mediaLink, file.path);
      const data = await this.app.vault.readBinary(dataFile);
      temp.push(storeMediaFile(path.basename(mediaLink), bytesToBase64(new Uint8Array(data))));
    }
 }
 requests.push(multi(temp));
 temp = [];
 this.requests_1_result = await invoke('multi', { actions: requests });
 await this.parse_requests_1();
...
```

I suppose that instead of grouping media per file, a solution would be to group them per note. Or to add media one file at a time.

EDIT: It isn't timing out! That was my own fault for closing Anki. But it is taking an unreasonably long time that might mislead users.

**Comment by Pseudonium** [comment]
Hmm, so it actually throws an error? The old python script used to take minutes for me to run properly.

Each HTTP request does itself take an amount of time - in the very early days of the python script (before it was a plugin), I initially set it up to do one or more requests per note. It ended up taking ages! That’s why I bundle the requests as much as possible.

I’ll see what I can do to fix this, if anything.

**Comment by Pseudonium** [comment]
Very odd... I just tested adding like 7 media files at once, and it didn't time out? Granted, it still took like 12 seconds, but no error was thrown...

Pictures:

![image](https://user-images.githubusercontent.com/27702145/102014089-20152e80-3d4c-11eb-9013-8ba9c6b1d5f3.png)
![image](https://user-images.githubusercontent.com/27702145/102014091-260b0f80-3d4c-11eb-9bf3-1091ded78110.png)
![image](https://user-images.githubusercontent.com/27702145/102014103-2e634a80-3d4c-11eb-9cc4-e3f80162689c.png)
![image](https://user-images.githubusercontent.com/27702145/102014110-34592b80-3d4c-11eb-97d6-7d8ff5aa2700.png)


And the network tab:

![image](https://user-images.githubusercontent.com/27702145/102014142-71252280-3d4c-11eb-9bee-7c1ecfcda0f8.png)

**Comment by Pseudonium** [comment]
And it's not like it should be an internet issue or anything, everything's local

**Comment by Pseudonium** [comment]
Another idea - at the moment, the way it works is that the plugin reads the binary contents of the file, encodes it in base 64, and then passes that string to AnkiConnect, which adds the media file. AnkiConnect does also provide another method of adding a media file - specifying the absolute path to the file. I'm not sure whether that would be faster? The other option is providing a URL to download the file from, which of course won't work for local files.

**Comment by jqhoogland** [comment]
I looked back and the timeout was not an actual timeout—I had closed Anki while I had left this running in the background.
Apologies for misleading (though this might prove a problem for other users who don't realize this takes so long).

Maybe you can add a comment to the wiki (or in the pop-up) that "This could take minutes. Don't close Anki in the meantime."

I'm scanning again and am now just waiting it out—I'll let you know how long it ends up taking.

I'll also try with the absolute path. But in any case it's weird that it's taking taking me minutes to complete even 5 or 6 files.

**Comment by Pseudonium** [comment]
I just tested it using the absolute path to the file rather than supplying the data myself, and it was much faster - down to half a second instead of 12 seconds! I'll see about integrating this into an upcoming update, hopefully it'll improve the speed for you.

**Comment by Pseudonium** [comment]
Should be fixed in v3.3.2, feel free to reopen the issue if the problem persists
